File: internal/telemetry/stats.go
package telemetry

import "sync/atomic"

type Stats struct {
	TotalKeys      atomic.Int64
	WALSegments    atomic.Int64
	SSTables       atomic.Int64
	LastWriteMS    atomic.Int64
	LastReadNS     atomic.Int64
	LastRecoveryMS atomic.Int64
	Compactions    atomic.Int64
	Flushes        atomic.Int64
}

func (s *Stats) Snapshot() Snapshot {
	return Snapshot{
		TotalKeys:      s.TotalKeys.Load(),
		WALSegments:    s.WALSegments.Load(),
		SSTables:       s.SSTables.Load(),
		LastWriteMS:    s.LastWriteMS.Load(),
		LastReadNS:     s.LastReadNS.Load(),
		LastRecoveryMS: s.LastRecoveryMS.Load(),
		Compactions:    s.Compactions.Load(),
		Flushes:        s.Flushes.Load(),
	}
}

File: internal/telemetry/stats.cpp
#include "stats.h"

namespace bigdb::telemetry {

Snapshot Stats::snapshot() const {
    Snapshot s;
    s.total_keys = total_keys.load(std::memory_order_relaxed);
    s.wal_segments = wal_segments.load(std::memory_order_relaxed);
    s.sstables = sstables.load(std::memory_order_relaxed);
    s.last_write_ms = last_write_ms.load(std::memory_order_relaxed);
    s.last_read_ns = last_read_ns.load(std::memory_order_relaxed);
    s.last_recovery_ms = last_recovery_ms.load(std::memory_order_relaxed);
    s.compactions = compactions.load(std::memory_order_relaxed);
    s.flushes = flushes.load(std::memory_order_relaxed);
    return s;
}

} // namespace bigdb::telemetry
